pkg/kube_events_manager/monitor_test.go

```diff
@@ -0,0 +1,143 @@
+)
+
+func Test_Monitor_should_handle_dynamic_ns_events(t *testing.T) {
+	g := NewWithT(t)
+	fc := fake.NewFakeCluster(fake.ClusterVersionV121)
+
+	// Initial namespace with ConfigMap.
+	createNsWithLabels(fc, "default", map[string]string{"test-label": ""})
+	createCM(fc, "default", testCM("default-cm-1"))
+
+	monitorCfg := &MonitorConfig{
+		ApiVersion: "v1",
+		Kind:       "ConfigMap",
+		EventTypes: []WatchEventType{WatchEventAdded, WatchEventModified, WatchEventDeleted},
+		NamespaceSelector: &NamespaceSelector{
+			LabelSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"test-label": "",
+				},
+			},
+		},
+	}
+
+	mon := NewMonitor()
+	mon.WithContext(context.TODO())
+	mon.WithKubeClient(fc.Client)
+	mon.WithConfig(monitorCfg)
+
+	// Catch KubeEvents from monitor.
+	type kubeEvent struct {
+		Type  KubeEventType
+		Count int
+	}
+	events := make([]kubeEvent, 0)
+	mon.WithKubeEventCb(func(ev KubeEvent) {
+		events = append(events, kubeEvent{
+			Type:  ev.Type,
+			Count: len(ev.Objects),
+		})
+	})
+
+	// Start monitor.
+	err := mon.CreateInformers()
+	g.Expect(err).ShouldNot(HaveOccurred())
+	mon.Start(context.TODO())
+
+	// Get initial snapshot.
+	snap := mon.Snapshot()
+	g.Expect(snap).Should(HaveLen(1), "Should have only one ConfigMap on start")
+	g.Expect(events).Should(HaveLen(0), "Should not call KubeEventCb until EnableKubeEventsCb")
+
+	// Simulate object creation during Synchronization phase:
+	// create new ns with matching labels and then create new ConfigMap.
+	createNsWithLabels(fc, "test-ns-1", map[string]string{"test-label": ""})
+
+	// Wait until informer appears.
+	g.Eventually(func() bool {
+		monImpl := mon.(*monitor)
+		return len(monImpl.VaryingInformers) == 2
+	}, "5s", "10ms").Should(BeTrue(), "Should create informer for new Namespace")
+
+	createCM(fc, "test-ns-1", testCM("cm-1"))
+
+	// Should update snapshot. Snapshot resets cached events.
+	g.Eventually(func() bool {
+		snap := mon.Snapshot()
+		return len(snap) == 2
+	}, "5s", "10ms").Should(BeTrue(), "Should not update snapshot until EnableKubeEventsCb")
+
+	// Create more ConfigMaps to cache some events.
+	createCM(fc, "test-ns-1", testCM("cm-2"))
+	createCM(fc, "test-ns-1", testCM("cm-3"))
+
+	g.Expect(events).Should(HaveLen(0), "Should not fire KubeEvents until EnableKubeEventsCb")
+
+	// Enable Kube events, should update snapshots now.
+	mon.EnableKubeEventCb()
+
+	// Should catch 2 events for cm-2 and cm-3.
+	g.Eventually(func() bool {
+		return len(events) == 2
+	}, "6s", "10ms").Should(BeTrue(), "Should fire cached KubeEvents after EnableKubeEventCb")
+
+	// Now simulate NS creation and objects creation after Synchronization phase.
+
+	// Create new ns with labels and cm there.
+	createNsWithLabels(fc, "test-ns-2", map[string]string{"test-label": ""})
+
+	// Wait until informer for new NS appears.
+	g.Eventually(func() bool {
+		monImpl := mon.(*monitor)
+		return len(monImpl.VaryingInformers) == 3
```

Comment:
Counting seems good enough since this behavior looks deterministic. But I suggest to be more strict and test that `VaryingInformers` actually contains key "test-ns-2"

Reply:
I've rewrite this test.